compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -26,6 +26,11 @@
 //! Here the block (`{ return; }`) has the return type `char`, rather than `()`, but the MIR we
 //! naively generate still contains the `_a = ()` write in the unreachable block "after" the
 //! return.
+//!
+//! **WARNING**: This is one of the few optimizations that runs on built and analysis MIR, and
+//! so its effects may affect the type-checking, borrow-checking, and other analysis of MIR.
+//! We must be extremely careful to only apply optimizations that preserve the semantics of the
+//! code, since changes here can affect which programs compile in an insta-stable way.
```

Comment:
Removing UB is generally considered "preserving the semantics", since UB programs have "all behaviors". So I think the wording here is potentially confusing.
```suggestion
//! We must be extremely careful to only apply optimizations that preserve UB and
//! all non-determinism, since changes here can affect which programs compile in
//! an insta-stable way. The normal logic that a program with UB can be changed
//! to do anything does not apply to pre-"runtime" MIR!
```